Sources/SKCore/TaskScheduler.swift

```diff
@@ -0,0 +1,459 @@
+            switch await task.value {
+            case .suspended:
+              // Break the switch and wait for a new `suspendableTask` to be placed into `suspendableTaskCreatedStream`.
+              break
+            case .terminated:
+              // The task finished. We are done with this `QueuedTask`
+              return
+            }
+          }
+        }
+        // The first (update priority) task never finishes, so this waits for the second (wait for suspendable) task
+        // to terminate.
+        // Afterwards we also cancel the update priority task.
+        for await _ in taskGroup {
+          taskGroup.cancelAll()
+          return
+        }
+      }
+    }
+  }
+
+  /// Start executing the task. Execution might be suspended, in which case this returns `.suspended`. In that case
+  /// the `TaskScheduler` is expected to call `execute` again.
+  func execute() async -> SuspendableTaskFinishStatus {
+    precondition(suspendableTask == nil, "Task started twice")
+    let task = Task.detached(priority: self.priority) {
+      await self.description.execute()
+      return await self.finalizeExecution()
+    }
+    suspendableTask = task
+    suspendableTaskCreatedContinuation.yield(task)
+    return await task.value
+  }
+
+  /// Implementation detail of `execute` that is called after `self.description.execute` finishes.
+  private func finalizeExecution() -> SuspendableTaskFinishStatus {
+    self.suspendableTask = nil
+    if Task.isCancelled && self.suspended.value {
+      self.suspended.value = false
+      return SuspendableTaskFinishStatus.suspended
+    } else {
+      return SuspendableTaskFinishStatus.terminated
+    }
+  }
+
+  /// Try suspending this task.
+  ///
+  /// Suspension might not do anything if the task has already finished, ie. even though `suspend` was called, the
+  /// task might still
+  func suspend() {
+    self.suspended.value = true
+    suspendableTask?.cancel()
+    self.suspendableTask = nil
+  }
+
+  /// Trigger `QueuedTask.priority` to be updated with the current priority of the underlying task.
+  ///
+  /// This is an asynchronous operation that makes no guarantees when the updated priority will be available.
+  ///
+  /// This is needed because tasks can't subscribe to priority updates (ie. there is no `withPriorityHandler` similar to
+  /// `withCancellationHandler`)
+  func triggerPriorityUpdate() {
+    updatePriorityContinuation.yield()
+  }
+
+  /// If the priority of this task is less than `targetPriority`, elevate the priority to `targetPriority` by spawning
+  /// a new task that depends on it. Otherwise a no-op.
+  nonisolated func elevatePriority(to targetPriority: TaskPriority) {
+    if priority < targetPriority {
+      logger.debug(
+        "Elevating priority of \(self.description.forLogging) from \(self.priority.rawValue) to \(targetPriority.rawValue)"
+      )
+      Task.detached(priority: targetPriority) {
```

Comment:
probably doesn't need to detach -- plain task with priority should be fine -- and get task locals copied then

Reply:
If I remember dispatch behavior this will not necessarily be 100% reliable... even if you passed highest priority it may decide to use lower I think... but the general idea is okey I suppose.

Please file an issue about an API for `__priorityBoost(task, priority)` even if underscored I think we should offer this API.

Reply:
Filed https://github.com/apple/swift/issues/73369